README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# tufte

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/tufte)](https://CRAN.R-project.org/package=tufte)
[![Downloads from the RStudio CRAN
mirror](https://cranlogs.r-pkg.org/badges/tufte)](https://cran.r-project.org/package=tufte)
[![R-CMD-check](https://github.com/rstudio/tufte/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/rstudio/tufte/actions/workflows/R-CMD-check.yaml)
[![Codecov test
coverage](https://codecov.io/gh/rstudio/tufte/branch/main/graph/badge.svg)](https://app.codecov.io/gh/rstudio/tufte?branch=main)
<!-- badges: end -->

This R package provides a few R Markdown output formats that use the
Tufte style. See <https://rstudio.github.io/tufte/> for a comprehensive
example.

## Books

<a href="https://bookdown.org/yihui/rmarkdown/tufte-handouts.html"><img src="https://bookdown.org/yihui/rmarkdown/images/cover.png" alt="R Markdown: The Definitive Guide" class="book" height="400"/></a>

See about the Tufte Handouts format in R Markdown Definitive Guide

## Installation

You can install the last available released version from
[CRAN](https://cran.r-project.org/package=tufte)

``` r
install.packages('tufte')
```

You can also install the development version of **tufte** from
[GitHub](https://github.com/) with:

``` r
# install.packages("pak")
pak::pak("rstudio/tufte")
```

## Usage

The easiest way to make a new R Markdown document using Tufte style is
from within RStudio. Go to *File \> New File \> R Markdown \> From
template \> Tufte Handout*.

This can also be created from the command line using

``` r
rmarkdown::draft("tufte.Rmd", "tufte_html", "tufte")
```

## Getting help

There are two main places to get help:

1.  The [RStudio
    community](https://community.rstudio.com/c/r-markdown/10) is a
    friendly place to ask any questions about rmarkdown and the R
    Markdown family of packages. Use tag **tufte** in your post.

2.  [Stack
    Overflow](https://stackoverflow.com/questions/tagged/r-markdown+tufte)
    is a great source of answers to common rmarkdown questions. It is
    also a great place to get help, once you have created a reproducible
    example that illustrates your problem.


NEWS.md
# CHANGES IN tufte VERSION 0.13

- Fix issue with margin content when using Pandoc 2.19 and later by adding `--wrap preserve` by default (thanks, @TomBen, #115).

# CHANGES IN tufte VERSION 0.12

- Fix footnotes as sidenotes issue with Pandoc 2.15 and later (thanks, @MCMaurer, #108).

# CHANGES IN tufte VERSION 0.11

- Fix an issue with Pandoc's table support in `tufte_handout()` (thanks, @chrisjake, #104).

- Fix an issue with Pandoc's citation processing by updating latex template (thanks, @rmcd1024, #106).

# CHANGES IN tufte VERSION 0.10

- Fix an issue with the wrong references placed in margins when `link-citations: yes` is used with Pandoc 2.11+ (#92)

# CHANGES IN tufte VERSION 0.9

- Add the missing `CSLReferences` environment in the template required by Pandoc 2.11+ and its new citeproc (#89)

- `tufte_handout()` now uses default `tidy` knitr option, which is `FALSE`. It can be changed with `knitr::opts_chunk$set(tidy = TRUE)` and requires in that case the **formatR** package.

# CHANGES IN tufte VERSION 0.8

- References are now moved in the margin correctly with Pandoc 2.11 (#86).

- Add a `runningheader` variable in the template for `tufte_handout()` and `tufte_book()` to have a different running header than the title if provided. (#82)

# CHANGES IN tufte VERSION 0.7

- Add the `cslreferences` environment in the template (thanks, @jonathan-g, #80).

# CHANGES IN tufte VERSION 0.6

## NEW FEATURES

- Arbitrary full width content is supported now; see https://rstudio.github.io/tufte/#arbitrary-full-width-content for the syntax (thanks, @mlovelace @josiekre, #33).

## BUG FIXES

- Footnotes no longer work with Pandoc >= 2.7 (thanks, @grimbough, #76).

- Margin references do not work with Pandoc 2.9.2.1 (thanks, @Selbosh, #78).

# CHANGES IN tufte VERSION 0.5

## BUG FIXES

- Make references generated by Pandoc >= 2.7 work in the margin (thanks, @dfeng, #64).

# CHANGES IN tufte VERSION 0.4

## BUG FIXES

- Should use the Pandoc argument `--top-level-division=chapter` instead of `--chapters` for `tufte_book()` (thanks, @jtth, #52).

- Processing of multi-line footnotes fails with the Pandoc option `--wrap preserve` (thanks, @aoles, #57 #58).

# CHANGES IN tufte VERSION 0.3

## NEW FEATURES

- It is possible to select a subset of some features of the default Tufte style (`tufte-css`) via the `tufte_features` argument of `tufte_html()`:

    - Disable the `et-book` fonts.
    - Remove the default light-yellow background color.
    - Use italics for document headers or not.

- A new variant of the Tufte style, `envisioned`, is added to `tufte_html()`. You can use `tufte_html(tufte_variant = 'envisioned')` to enable this style. The major difference with the default Tufte style is: the font family is `Roboto Condensed`, the background color is `#fefefe`, and the text color is `#222` (thanks, @eddelbuettel, #21).

- You can choose whether references from citations should be placed in the document margins or at the bottom using the `margin_references` argument of `tufte_html()` (thanks, @stefanfritsch, #49).

## BUG FIXES

- When `link-citations: no` in YAML, citations should not be moved into the page margin in the HTML output (https://stackoverflow.com/q/39053097/559676).

- Horizontal lines could bisect margin notes and footnotes (thanks, @ajdamico, #32).

- Compatibility issues with Pandoc 2.0 (thanks, @peetCreative, #51).

# CHANGES IN tufte VERSION 0.2

## NEW FEATURES

- it is possible to generate PDF output using the ctex package for R Markdown documents in Chinese; you just need to specify `ctex: yes` in the YAML metadata

## BUG FIXES

- the default LaTeX template for `tufte_handout()` may not work when the LaTeX
  package **soul** is installed, or **ifxetex**/**xltxtra** are not installed

- the `number_sections` option did not work for LaTeX output

# CHANGES IN tufte VERSION 0.1

## NEW FEATURES

- added three output formats for R Markdown: `tufte_html()`, `tufte_handout()`, 
and `tufte_book()`


tests/testthat.R
library(testthat)
library(tufte)

test_check("tufte")


tests/testthat/helpers.R
# Use to test pandoc availability or version lower than
skip_if_not_pandoc <- function(ver = NULL) {
  if (!rmarkdown::pandoc_available(ver)) {
    msg <- if (is.null(ver)) {
      "Pandoc is not available"
    } else {
      sprintf("Version of Pandoc is lower than %s.", ver)
    }
    skip(msg)
  }
}

# Use to test version greater than
skip_if_pandoc <- function(ver = NULL) {
  if (rmarkdown::pandoc_available(ver)) {
    msg <- if (is.null(ver)) {
      "Pandoc is available"
    } else {
      sprintf("Version of Pandoc is greater than %s.", ver)
    }
    skip(msg)
  }
}

local_rmd_file <- function(..., .env = parent.frame()) {
  path <- withr::local_tempfile(.local_envir = .env, fileext = ".Rmd")
  xfun::write_utf8(c(...), path)
  path
}

local_render <- function(input, ..., .env = parent.frame()) {
  skip_if_not_pandoc()
  output_file <- withr::local_tempfile(.local_envir = .env)
  rmarkdown::render(input, output_file = output_file, quiet = TRUE, ...)
}

local_pandoc_convert <- function(text, from = "markdown", options = NULL, ..., .env = parent.frame()) {
  skip_if_not_pandoc()
  rmd <- local_rmd_file(text)
  out <- withr::local_tempfile(.local_envir = .env)
  rmarkdown::pandoc_convert(rmd,
    from = from, output = out,
    options = c("--wrap", "preserve", options), ...
  )
  xfun::read_utf8(out)
}

.render_and_read <- function(input, ...) {
  skip_if_not_pandoc()
  res <- local_render(input, ...)
  xfun::read_utf8(res)
}


R/utils.R
.onLoad <- function(lib, pkg) {
  # this engine will be overridden in tufte_html/tufte_handout formats
  knitr::knit_engines$set(marginfigure = function(options) {
    'Placeholder (you should not see this)'
  })
}

#' @details `newthought()` can be used in inline R expressions in R
#'   Markdown
#'   ```r
#'   `r newthought(Some text)`
#'   ```
#'   and it works for both
#'   HTML (\samp{<span class="newthought">text</span>}) and PDF
#'   (\samp{\\newthought{text}}) output.
#' @param text A character string to be presented as a \dQuote{new thought}
#'   (using small caps), or a margin note, or a footer of a quote
#' @rdname tufte_handout
#' @export
#' @examples newthought("In this section")
newthought <- function(text) {
  if (is_html_output()) {
    sprintf('<span class="newthought">%s</span>', text)
  } else if (is_latex_output()) {
    sprintf("\\newthought{%s}", text)
  } else {
    sprintf('<span style="font-variant:small-caps;">%s</span>', text)
  }
}

#' @details `margin_note()` can be used in inline R expressions to write a
#'   margin note (like a sidenote but not numbered).
#' @param icon A character string to indicate there is a hidden margin note when
#'   the page width is too narrow (by default it is a circled plus sign)
#' @rdname tufte_handout
#' @importFrom knitr is_html_output is_latex_output
#' @export
margin_note <- function(text, icon = "&#8853;") {
  if (is_html_output()) {
    marginnote_html(sprintf('<span class="marginnote">%s</span>', text), icon)
  } else if (is_latex_output()) {
    sprintf("\\marginnote{%s}", text)
  } else {
    warning("marginnote() only works for HTML and LaTeX output", call. = FALSE)
    text
  }
}

#' @details `quote_footer()` formats text as the footer of a quote. It puts
#'   `text` in \samp{<footer></footer>} for HTML output, and
#'   after \samp{\\hfill} for LaTeX output (to right-align text).
#' @rdname tufte_handout
#' @export
quote_footer <- function(text) {
  if (is_html_output()) {
    sprintf("<footer>%s</footer>", text)
  } else if (is_latex_output()) {
    sprintf("\\hfill %s", text)
  } else {
    warning("quote_footer() only works for HTML and LaTeX output", call. = FALSE)
    text
  }
}

#' @details `sans_serif()` applies sans-serif fonts to `text`.
#' @rdname tufte_handout
#' @export
sans_serif <- function(text) {
  if (is_html_output()) {
    sprintf('<span class="sans">%s</span>', text)
  } else if (is_latex_output()) {
    sprintf("\\textsf{%s}", text)
  } else {
    warning("sans_serif() only works for HTML and LaTeX output", call. = FALSE)
    text
  }
}

template_resources <- function(name, ...) {
  system.file("rmarkdown", "templates", name, "resources", ..., package = "tufte")
}

gsub_fixed <- function(...) gsub(..., fixed = TRUE)

pandoc2.0 <- function() rmarkdown::pandoc_available("2.0")

# add --wrap=preserve to pandoc args for pandoc 2.0:
# https://github.com/rstudio/bookdown/issues/504
# https://github.com/rstudio/tufte/issues/115
add_wrap_preserve <- function(args, pandoc2 = pandoc2.0) {
  if (pandoc2 && !length(grep("--wrap", args))) {
    c("--wrap", "preserve", args)
  } else {
    args
  }
}


R/handout.R
#' Tufte handout formats (PDF and HTML)
#'
#' Templates for creating handouts according to the style of Edward R. Tufte and
#' Richard Feynman.
#'
#' `tufte_handout()` provides the PDF format based on the Tufte-LaTeX
#' class: <https://tufte-latex.github.io/tufte-latex/>.
#' @inheritParams rmarkdown::pdf_document
#' @param ... Other arguments to be passed to [pdf_document()] or
#'   [html_document()] (note you cannot use the `template`
#'   argument in `tufte_handout` or the `theme` argument in
#'   `tufte_html()`; these arguments have been set internally)
#' @references See <https://rstudio.github.io/tufte/> for an example.
#' @export
#' @examples library(tufte)
tufte_handout <- function(fig_width = 4, fig_height = 2.5, fig_crop = TRUE, dev = "pdf",
                          highlight = "default", ...) {
  tufte_pdf("tufte-handout", fig_width, fig_height, fig_crop, dev, highlight, ...)
}

#' @rdname tufte_handout
#' @export
tufte_book <- function(fig_width = 4, fig_height = 2.5, fig_crop = TRUE, dev = "pdf",
                       highlight = "default", ...) {
  tufte_pdf("tufte-book", fig_width, fig_height, fig_crop, dev, highlight, ...)
}

tufte_pdf <- function(documentclass = c("tufte-handout", "tufte-book"), fig_width = 4, fig_height = 2.5,
                      fig_crop = TRUE, dev = "pdf", highlight = "default",
                      template = template_resources("tufte_handout", "tufte-handout.tex"), ...) {

  # resolve default highlight
  if (identical(highlight, "default")) highlight <- "pygments"

  # call the base pdf_document format with the appropriate options
  format <- rmarkdown::pdf_document(
    fig_width = fig_width, fig_height = fig_height, fig_crop = fig_crop,
    dev = dev, highlight = highlight, template = template, ...
  )

  # LaTeX document class
  documentclass <- match.arg(documentclass)
  format$pandoc$args <- c(
    format$pandoc$args, "--variable", paste0("documentclass:", documentclass),
    if (documentclass == "tufte-book") {
      if (pandoc2.0()) "--top-level-division=chapter" else "--chapters"
    }
  )

  knitr::knit_engines$set(marginfigure = function(options) {
    options$type <- "marginfigure"
    eng_block <- knitr::knit_engines$get("block")
    eng_block(options)
  })

  # create knitr options (ensure opts and hooks are non-null)
  knitr_options <- rmarkdown::knitr_options_pdf(fig_width, fig_height, fig_crop, dev)
  if (is.null(knitr_options$opts_knit)) knitr_options$opts_knit <- list()
  if (is.null(knitr_options$knit_hooks)) knitr_options$knit_hooks <- list()

  # set options
  knitr_options$opts_knit$width <- 45

  # set hooks for special plot output
  knitr_options$knit_hooks$plot <- function(x, options) {

    # determine figure type
    if (isTRUE(options$fig.margin)) {
      options$fig.env <- "marginfigure"
      if (is.null(options$fig.cap)) options$fig.cap <- ""
    } else if (isTRUE(options$fig.fullwidth)) {
      options$fig.env <- "figure*"
      if (is.null(options$fig.cap)) options$fig.cap <- ""
    }

    knitr::hook_plot_tex(x, options)
  }

  # override the knitr settings of the base format and return the format
  format$knitr <- knitr_options
  format$inherits <- "pdf_document"
  format
}


R/html.R
#' @details `tufte_html()` provides the HTML format based on the Tufte CSS:
#'   <https://edwardtufte.github.io/tufte-css/>.
#' @param tufte_features A character vector of style features to enable:
#'   `fonts` stands for the `et-book` fonts in the `tufte-css`
#'   project, `background` means the lightyellow background color of the
#'   page, and `italics` means whether to use italics for the headers. You
#'   can enable a subset of these features, or just disable all of them by
#'   `NULL`. When this argument is not used and the `tufte_variant`
#'   argument is not `default`, no features are enabled.
#' @param tufte_variant A variant of the Tufte style. Currently supported styles
#'   are `default` (from the `tufte-css` project), and
#'   `envisioned` (inspired by the project `Envisioned CSS`
#'   <https://github.com/nogginfuel/envisioned-css> but essentially just
#'   sets the font family to `Roboto Condensed`, and changed the
#'   background/foreground colors).
#' @param margin_references Whether to place citations in margin notes.
#' @rdname tufte_handout
#' @export
tufte_html <- function(..., tufte_features = c("fonts", "background", "italics"),
                       tufte_variant = c("default", "envisioned"), margin_references = TRUE) {
  tufte_variant <- match.arg(tufte_variant)
  if (missing(tufte_features) && tufte_variant != "default") {
    tufte_features <- character()
  }
  html_document2 <- function(..., extra_dependencies = list()) {
    rmarkdown::html_document(
      ...,
      extra_dependencies = c(
        extra_dependencies, tufte_html_dependency(tufte_features, tufte_variant)
      )
    )
  }
  format <- html_document2(theme = NULL, ...)
  pandoc2 <- pandoc2.0()

  # add --wrap=preserve to pandoc args for pandoc 2.0:
  format$pandoc$args <-  add_wrap_preserve(format$pandoc$args, pandoc2)

  # when fig.margin = TRUE, set fig.beforecode = TRUE so plots are moved before
  # code blocks, and they can be top-aligned
  ohooks <- knitr::opts_hooks$set(fig.margin = function(options) {
    if (isTRUE(options$fig.margin)) {
      options$fig.beforecode <- TRUE
    }
    options
  })

  # make sure the existing post processor is called first in our new processor
  post_processor <- format$post_processor
  format$post_processor <- function(metadata, input, output, clean, verbose) {
    if (is.function(post_processor)) {
      output <- post_processor(metadata, input, output, clean, verbose)
    }

    knitr::opts_hooks$restore(ohooks)

    x <- xfun::read_utf8(output)
    fn_label <- paste0(knitr::opts_knit$get("rmarkdown.pandoc.id_prefix"), "fn")
    footnotes <- parse_footnotes(x, fn_label)
    notes <- footnotes$items
    # replace footnotes with sidenotes
    for (i in seq_along(notes)) {
      num <- sprintf(
        '<a href="#%s%d" class="%s" id="%sref%d"><sup>%d</sup></a>',
        fn_label, i, if (pandoc2) "footnote-ref" else "footnoteRef", fn_label, i, i
      )
      con <- sprintf(paste0(
        '<label for="tufte-sn-%d" class="margin-toggle sidenote-number">%d</label>',
        '<input type="checkbox" id="tufte-sn-%d" class="margin-toggle">',
        '<span class="sidenote"><span class="sidenote-number">%d</span> %s</span>'
      ), i, i, i, i, notes[i])
      x <- gsub_fixed(num, con, x)
    }
    # remove footnotes at the bottom
    if (length(footnotes$range)) x <- x[-footnotes$range]

    # replace citations with margin notes
    if (margin_references) x <- margin_references(x)

    # place figure captions in margin notes
    x[x == '<p class="caption">'] <- '<p class="caption marginnote shownote">'

    # move </caption> to the same line as <caption>; the previous line should
    # start with <table
    for (i in intersect(grep("^<caption>", x), grep("^<table", x) + 1)) {
      j <- 0
      while (!grepl("</caption>$", x[i])) {
        j <- j + 1
        x[i] <- paste0(x[i], x[i + j])
        x[i + j] <- ""
      }
    }
    # place table captions in the margin
    r <- "^<caption>(.+)</caption>$"
    for (i in grep(r, x)) {
      # the previous line should be <table> or <table class=...>
      if (!grepl("^<table( class=.+)?>$", x[i - 1])) next
      cap <- gsub(r, "\\1", x[i])
      x[i] <- x[i - 1]
      x[i - 1] <- paste0(
        "<p><!--\n<caption>-->", '<span class="marginnote shownote">',
        cap, "</span><!--</caption>--></p>"
      )
    }

    # add an incremental number to the id of <label> and <input> for margin notes
    r <- '(<label|<input type="checkbox") (id|for)(="tufte-mn)-(" )'
    m <- gregexpr(r, x)
    j <- 1
    regmatches(x, m) <- lapply(regmatches(x, m), function(z) {
      n <- length(z)
      if (n == 0) {
        return(z)
      }
      if (n %% 2 != 0) warning("The number of labels is different with checkboxes")
      for (i in seq(1, n, 2)) {
        if (i + 1 > n) break
        z[i + (0:1)] <- gsub(r, paste0("\\1 \\2\\3-", j, "\\4"), z[i + (0:1)])
        j <<- j + 1
      }
      z
    })

    xfun::write_utf8(x, output)
    output
  }

  if (is.null(format$knitr$knit_hooks)) format$knitr$knit_hooks <- list()
  format$knitr$knit_hooks$plot <- function(x, options) {
    # make sure the plot hook always generates HTML code instead of ![]()
    if (is.null(options$out.extra)) options$out.extra <- ""
    fig_margin <- isTRUE(options$fig.margin)
    fig_fullwd <- isTRUE(options$fig.fullwidth)
    if (fig_margin || fig_fullwd) {
      if (is.null(options$fig.cap)) options$fig.cap <- " " # empty caption
    } else if (is.null(options$fig.topcaption)) {
      # for normal figures, place captions at the top of images
      options$fig.topcaption <- TRUE
    }
    res <- knitr::hook_plot_md(x, options)
    if (fig_margin) {
      res <- gsub_fixed('<p class="caption">', '<!--\n<p class="caption marginnote">-->', res)
      res <- gsub_fixed("</p>", "<!--</p>-->", res)
      res <- gsub_fixed("</div>", "<!--</div>--></span></p>", res)
      res <- gsub_fixed(
        '<div class="figure">', paste0(
          "<p>", '<span class="marginnote shownote">', '\n<!--\n<div class="figure">-->'
        ), res
      )
    } else if (fig_fullwd) {
      res <- gsub_fixed('<div class="figure">', '<div class="figure fullwidth">', res)
      res <- gsub_fixed(
        '<p class="caption">', '<p class="caption marginnote shownote">', res
      )
    }
    res
  }

  knitr::knit_engines$set(marginfigure = function(options) {
    options$type <- "marginnote"
    if (is.null(options$html.tag)) options$html.tag <- "span"
    options$html.before <- marginnote_html()
    eng_block <- knitr::knit_engines$get("block")
    eng_block(options)
  })

  format$inherits <- "html_document"

  format
}

#' @importFrom htmltools htmlDependency
tufte_html_dependency <- function(features, variant) {
  list(htmlDependency(
    "tufte-css", "2015.12.29",
    src = template_resources("tufte_html"), stylesheet = c(
      sprintf("tufte-%s.css", features), "tufte.css",
      if (variant != "default") sprintf("%s.css", variant)
    )
  ))
}

# we assume one footnote only contains one paragraph here, although it is
# possible to write multiple paragraphs in a footnote with Pandoc's Markdown
parse_footnotes <- function(x, fn_label = "fn") {
  i <- grep('^<div class="footnotes[^"]*"[^>]*>', x)
  if (length(i) == 0) {
    return(list(items = character(), range = integer()))
  }
  j <- which(x == "</div>")
  j <- min(j[j > i])
  n <- length(x)
  r <- sprintf(
    '<li id="%s([0-9]+)"><p>(.+)<a href="#%sref\\1"([^>]*)>.{1,2}</a></p></li>',
    fn_label, fn_label
  )
  s <- paste(x[i:j], collapse = "\n")
  list(
    items = gsub(r, "\\2", unlist(regmatches(s, gregexpr(r, s)))),
    range = i:j
  )
}

# move reference items from the bottom to the margin (as margin notes)
margin_references <- function(x) {
  i <- grep('^<div id="refs" class="references[^"]*"[^>]*>$', x)
  if (length(i) != 1) {
    return(x)
  }
  # link-citations: no
  if (length(grep('<a href="#ref-[^"]+"[^>]*>([^<]+)</a>', x)) == 0) {
    return(x)
  }
  r <- '^<div id="(ref-[^"]+)"[^>]*>$'
  k <- grep(r, x)
  k <- k[k > i]
  n <- length(k)
  if (n == 0) {
    return(x)
  }
  # pandoc-citeproc may generate a link on both the year and the alphabetic
  # suffix, e.g. <a href="#cite-key">2016</a><a href="#cite-key">a</a>;
  # Pandoc 2.11+ may also generate several links with parenthesis between links,
  # but this is different, e.g
  # <a href="#ref-R-rmarkdown" role="doc-biblioref">Allaire et al.</a> (<a href="#ref-R-rmarkdown" role="doc-biblioref">2020</a>)</span>
  # we need to merge the two links
  r2 <- '(<a href="#ref-[^"]+"[^>]*>)([^<]+)</a>([^<]*)\\1([^<]+)</a>([^<]*)'
  x <- gsub(r2, "\\1\\2\\3\\4\\5</a>", x)
  ids <- gsub(r, "\\1", x[k])
  ids <- sprintf('<a href="#%s"[^>]*>([^<]+)</a>', ids)
  ref <- gsub("^<p>|</p>$", "", x[k + 1])
  # replace 3 em-dashes with author names
  dashes <- paste0("^", intToUtf8(rep(8212, 3)), "[.]")
  for (j in grep(dashes, ref)) {
    ref[j] <- sub(dashes, sub("^([^.]+[.])( .+)$", "\\1", ref[j - 1]), ref[j])
  }
  ref <- marginnote_html(paste0('\\1<span class="marginnote">', ref, "</span>"))
  for (j in seq_len(n)) {
    x <- gsub(ids[j], ref[j], x)
  }
  x[-(i:(max(k) + 3))] # remove references at the bottom
}

marginnote_html <- function(text = "", icon = "&#8853;") {
  sprintf(paste0(
    '<label for="tufte-mn-" class="margin-toggle">%s</label>',
    '<input type="checkbox" id="tufte-mn-" class="margin-toggle">%s'
  ), icon, text)
}
